### PR TITLE
[10.0][Add] sale_manual_delivery

### PR DESCRIPTION
**Pull request by leemannd**
Sale Manual Delivery
====================
The Goal of this module is to allow the manual creation of delivery slips. If activated,
a sale order won't direcly impact directly the stock. It will not make a reservation on the stock.
It allows the delivery and the impact on stock to be done manually when needed.
The goal is to be used on long term projects where not all the material is shipped at once.

### CODE REVIEW

File: sale_manual_delivery/wizard/line.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Denis Leemann, Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields
+
+
+class ManualProcurement(models.TransientModel):
```
Comment (yvaucher) [inline sale_manual_delivery/wizard/line.py]: ManualLine

File: sale_manual_delivery/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+Sale Manual Delivery
```
Comment (elicoidal) [inline sale_manual_delivery/README.md]: @leemannd   Thanks for your contribution to the OCA. 
Let me introduce to you some basic information about contribution in OCA and I will review other points in more comments inside the code.
In general, please follow the principles: 
* Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
* Please follow the OCA README template:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
* We recommend to include a screenshot as well to help new comers (nice to have)
* Please use simple headers in all py files following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

Comment (elicoidal) [reply sale_manual_delivery/README.md]: in particular here: you need to add a line as document title

File: sale_manual_delivery/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+====================
+
+The Goal of this module is to allow the manual creation of delivery slips. If activated,
```
Comment (elicoidal) [inline sale_manual_delivery/README.md]: s/Goal/goal

File: sale_manual_delivery/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+a sale order won't direcly impact directly the stock. It will not make a reservation on the stock.
+It allows the delivery and the impact on stock to be done manually when needed.
+The goal is to be used on long term projects where not all the material is shipped at once.
+
+Usage
+=====
```
Comment (elicoidal) [inline sale_manual_delivery/README.md]: remove the 2 empty sections

File: sale_manual_delivery/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+======================
+
+* WIP
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/sale-workflow/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
```
Comment (elicoidal) [inline sale_manual_delivery/README.md]: Update according newest template

File: sale_manual_delivery/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (elicoidal) [inline sale_manual_delivery/README.md]: https

File: sale_manual_delivery/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline sale_manual_delivery/__manifest__.py:1]: please add short headers

File: sale_manual_delivery/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+{
+    'name': 'Sale Manual Delivery',
+    'category': 'Sale',
+    'author': 'Camptocamp SA, Odoo Community Association (OCA)',
+    'version': '10.0.1.0.0',
+    'sequence': 150,
```
Comment (elicoidal) [inline sale_manual_delivery/__manifest__.py]: is it necessary?

File: sale_manual_delivery/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        'delivery',
+        'sale',
+        'sales_team',
+    ],
+    'data': [
+        'views/crm_team_view.xml',
+        'views/sale_order_view.xml',
+        'wizard/manual_proc_view.xml',
+    ],
+    'demo': [
```
Comment (elicoidal) [inline sale_manual_delivery/__manifest__.py]: remove empty key

File: sale_manual_delivery/models/procurement.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Denis Leemann, Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import models, fields
+
+
+class ProcurementOrder(models.Model):
+    _inherit = 'procurement.order'
+
+    carrier_id = fields.Many2one(
```
Comment (elicoidal) [inline sale_manual_delivery/models/procurement.py:10]: this looks irrelevant in the current context (module for manual picking, no carrier id)

**Review by elicoidal — CHANGES_REQUESTED** [review]
Light review about format as the module is on WIP process

File: sale_manual_delivery/models/procurement.py (hunk 10 above)
Comment (leemannd) [reply sale_manual_delivery/models/procurement.py:10]: @elicoidal Thanks for the feedbacks. One of the goals of this manual delivery, is to separate the delivery from the sale. The sale order is made once in the beginning of the project. Then parts are shipped later and be specified with another carrier.

Comment (elicoidal) [reply sale_manual_delivery/models/procurement.py:10]: @leemannd Then maybe you can add it to the module README to understand it better?

**Comment by jgrandguillaume** [comment]
Hi,

I've functionally reviewed this module and made some improvements on the general usage, tests and flow. It has now a smooth way of dealing with manual delivery creation. It inform the user of the already created delivery and controls that he makes no mistakes (e.g. ship more than ordered).

On my side, I'm :+1: 

Regards,

Joël

**Review by jgrandguillaume — APPROVED** [review]
Functional review ok

**Comment by JordiBForgeFlow** [comment]
@leemannd very nice and useful. Would be great to have it also in purchases. Only when the supplier confirms what is incoming will users create the appropiate incoming shipments , to match whst the supplier confirms as being shipped.

**Comment by leemannd** [comment]
@elicoidal It has been updated

File: sale_manual_delivery/README.rst [hunk 11]
```diff
@@ -0,0 +1,59 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=====================
+Sale Manual Delivery
+=====================
+
+The goal of this module is to allow the manual creation of delivery slips. If activated,
+a sale order won't direcly impact directly the stock. It will not make a reservation on the stock.
```
Comment (elicoidal) [inline sale_manual_delivery/README.rst:10]: directly is duplicated

**Review by elicoidal — APPROVED** [review]

File: sale_manual_delivery/README.rst [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+It allows the delivery and the impact on stock to be done manually when needed.
+The goal is to be used on long term projects where not all the material is shipped at once.
+As you make many shipments, every time you make a new delivery, you can choose an new carrier as well
+as a planned date. A wizard helps you to chose what to deliver by showing you how much you already 
+planned to ship.
+
+
```
Comment (aheficent) [inline sale_manual_delivery/README.rst]: I miss the usage section here

File: sale_manual_delivery/wizard/manual_proc_view.xml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+                  type='object'/>
+          <button string="Cancel" class="btn-default" special="cancel" />
+        </footer>
+      </form>
+
+    </field>
+  </record>
+
+  <act_window id='action_wizard_manual_delivery'
+              name='Create Manual Delivery'
```
Comment (AaronHForgeFlow) [inline sale_manual_delivery/wizard/manual_proc_view.xml:47]: I'm not sure if this is good. I'm getting a singleton error when I select more than one SO in the list view and then I click on the action.
image

**Review by AaronHForgeFlow — COMMENTED** [review]
Edited: dismissed comment

File: sale_manual_delivery/wizard/manual_proc_view.xml (hunk 13 above)
Comment (leemannd) [reply sale_manual_delivery/wizard/manual_proc_view.xml:47]: @aheficent Hello, Thank you for your feedback.
The wizard was intended to be launched from the SO and not from many. Currently I don't have the time to make improvements to this PR, but you can submit your idea. I'll be happy to review and include your commits into this PR.

Comment (AaronHForgeFlow) [reply sale_manual_delivery/wizard/manual_proc_view.xml:47]: Hello @leemannd sorry for the late reply. I'll create a PR to your branch then.

**Comment by yvaucher** [comment]
2 errors in tests:

```
FAIL: test_00_sale_manual_delivery (odoo.addons.sale_manual_delivery.tests.test_manual_delivery.TestSaleStock)
` Test SO's manual delivery
Traceback (most recent call last):
`   File "/home/travis/build/OCA/sale-workflow/sale_manual_delivery/tests/test_manual_delivery.py", line 53, in test_00_sale_manual_delivery
`     should be created after "manual delivery" wizard call')
` AssertionError: stock.picking() is not true : Sale Manual Delivery: picking             should be created after "manual delivery" wizard call
FAIL: test_02_sale_various_manual_delivery (odoo.addons.sale_manual_delivery.tests.test_manual_delivery.TestSaleStock)
` Test SO's various manual delivery
Traceback (most recent call last):
`   File "/home/travis/build/OCA/sale-workflow/sale_manual_delivery/tests/test_manual_delivery.py", line 165, in test_02_sale_various_manual_delivery
`     should be created after "manual delivery" wizard call')
` AssertionError: stock.picking() is not true : Sale Manual Delivery: picking             should be created after "manual delivery" wizard call
FAILED
Module sale_manual_delivery: 2 failures, 0 errors
```

**Comment by AaronHForgeFlow** [comment]
I just tested this module again and I think it works ok. Some minor issues found but can be easily fixed:

- Got error when trying to create the delivery for more than one SO. A warning message to the user telling that the operation is not allowed should be ok for me.

- If I cancel the SO and then I set it back to draft and then confirm the SO again. Then, when I create the deliveries those are not linked to the sale order, but they are actually created.

- The commented CI thing

@leemannd  I think it is worth it to move forward with this Pull Request. I will try to fix it if are good with that.

Regards.

**Comment by leemannd** [comment]
@aheficent Thanks for the review and time taken. I had currently no time allowed to work on it :(
IMO we should close this in favor of your PR

**Pull request closed without merging**